eval/topdown.go

```diff
@@ -65,89 +64,125 @@ func (ctx *TopDownContext) Bind(variable opalog.Var, value opalog.Value) *TopDow
 		return nil
 	}
 	cpy := *ctx
-	cpy.Bindings = make(Bindings)
-	for k, v := range ctx.Bindings {
-		cpy.Bindings[k] = plugValue(v, Bindings{variable: value})
-	}
-	cpy.Bindings[variable] = value
+	cpy.Bindings = newHashMap()
+	tmp := newHashMap()
+	tmp.Put(variable, value)
+	ctx.Bindings.Iter(func(k opalog.Value, v opalog.Value) bool {
+		cpy.Bindings.Put(k, plugValue(v, tmp))
+		return false
+	})
+	cpy.Bindings.Put(variable, value)
 	return &cpy
 }
 
+// Child returns a new context to evaluate a rule that was referenced by this context.
+func (ctx *TopDownContext) Child(rule *opalog.Rule, bindings *hashMap) *TopDownContext {
```

Comment:
At first I thought 'child' was an odd choice for the function, but on second thought maybe it's good b/c it's generating a TopDownContext whose parent is the original.